File: core/src/main/java/com/android/local/service/core/service/ALSService.java
package com.android.local.service.core.service;

import android.content.res.AssetManager;
import android.util.Log;

import com.android.local.service.core.ALSHelper;
import com.android.local.service.core.e.CustomResponse;
import com.android.local.service.core.i.RequestListener;

import org.apache.commons.fileupload.FileItem;
import org.apache.commons.fileupload.FileUpload;
import org.apache.commons.fileupload.RequestContext;
import org.apache.commons.fileupload.disk.DiskFileItemFactory;
import org.nanohttpd.protocols.http.IHTTPSession;
import org.nanohttpd.protocols.http.NanoHTTPD;
import org.nanohttpd.protocols.http.request.Method;
import org.nanohttpd.protocols.http.response.Response;
import org.nanohttpd.protocols.http.response.Status;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public class ALSService extends NanoHTTPD {

    private static final String TAG = ALSService.class.getSimpleName();
    // 统一响应字段
    public static String CODE_NAME = "code";
    public static String MESSAGE_NAME = "message";
    public static String DATA_NAME = "data";
    private RequestListener requestListener;
    private FileUpload upload;

    public ALSService(int port) {
        super(port);
    }

    private void initUpFile() {
        DiskFileItemFactory factory = new DiskFileItemFactory();
        upload = new FileUpload(factory);
    }

    public void setRequestListener(RequestListener listener) {
        this.requestListener = listener;
    }

    @Override
    @Deprecated
    public Response serve(IHTTPSession session) {
        Method method = session.getMethod();
        String uri = session.getUri();
        Map<String, String> params = session.getParms();

//        session.getParameters();
//        session.getCookies();
//        session.getInputStream();
//        session.getRemoteIpAddress();
//        session.getQueryParameterString();

        String contentType = "text/plain";
        String tmpType = session.getHeaders().get("content-type");
        if (tmpType == null) {
            tmpType = session.getHeaders().get("Content-Type");
        }
        if (tmpType == null) {
            tmpType = session.getHeaders().get("CONTENT-TYPE");
        }
        if (tmpType != null) {
            contentType = tmpType.toLowerCase(Locale.getDefault());
        }
        // 接收请求类型
        if (Method.POST.equals(method) || Method.PUT.equals(method) || Method.NOTIFY.equals(method)) {
            try {
                if (contentType.contains("multipart/form-data")) {
                    params = handleMultipartData(session);
                } else if (contentType.contains("application/json")) {//请求体中的数据是 JSON 格式
//                    String body = getRequestBody(session);
//                    Log.d("JSON Body", body);
//                    params = new HashMap<>();
//                    params.put("json", body);

                    Map<String, String> body = new HashMap<>();
                    session.parseBody(body);
                    Map<String, String> newParams = new HashMap<>();
                    newParams.put("json", body.get("postData"));
                    params = newParams;
                } else if (contentType.contains("/xml")) {//请求体中的数据是 XML 格式
//                    String body = getRequestBody(session);
//                    Log.d("XML Body", body);
//                    params = new HashMap<>();
//                    params.put("xml", body);

                    Map<String, String> body = new HashMap<>();
                    session.parseBody(body);
                    Map<String, String> newParams = new HashMap<>();
                    newParams.put("xml", body.get("postData"));
                    params = newParams;
                } else {
                    Map<String, String> body = new HashMap<>();
                    session.parseBody(body);
                    params = session.getParms();
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        String action = uri != null ? uri.substring(1) : null;
        if (action == null) {
            return wrapResponse(session, fileNotFoundResponse());
        }

        Log.d(TAG, "uri = " + uri + "   method = " + method + "   params = " + params);

        Response response;
        if (requestListener != null) {
            Map<String, String> headerMap = session.getHeaders();
            Map<String, String> bodyMap = params != null ? params : new HashMap<>();
            response = requestListener.onRequest(action, contentType, headerMap, bodyMap);
        } else {
            throw new RuntimeException("setRequestListener 方法没有设置");
        }

        return wrapResponse(session, response);
    }

    private String getRequestBody(IHTTPSession session) {
        int contentLength = 0;
        try {
            if (session.getHeaders().containsKey("content-length")) {
                String contentLengthStr = session.getHeaders().get("content-length");
                if (contentLengthStr != null) {
                    contentLength = Integer.parseInt(contentLengthStr);
                }
            }
        } catch (NumberFormatException e) {
            e.printStackTrace();
        }
        if (contentLength <= 0) {
            return "";
        }
        byte[] buffer = new byte[contentLength];
        try {
            int r = session.getInputStream().read(buffer, 0, contentLength);
            if (r <= 0) {
                return "";
            }
            return new String(buffer, "UTF-8");
        } catch (Exception e) {
            e.printStackTrace();
            return "";
        }
    }

    private Map<String, String> handleMultipartData(IHTTPSession session) {
        Map<String, String> result = new HashMap<>();
        Map<String, List<Map<String, Object>>> filesInfo = new HashMap<>();
        Map<String, String> formFields = new HashMap<>();

        try {
            RequestContext request = new RequestContext() {
                @Override
                public String getCharacterEncoding() {
                    String contentTypeHeader = session.getHeaders().get("content-type");
                    if (contentTypeHeader != null && contentTypeHeader.contains(";")) {
                        String[] parts = contentTypeHeader.trim().split(";");
                        if (parts.length > 1 && parts[1].contains("=")) {
                            return parts[1].split("=")[1];
                        }
                    }
                    return "UTF-8";
                }

                @Override
                public String getContentType() {
                    return session.getHeaders().get("content-type") != null ?
                            session.getHeaders().get("content-type") : "text/plain";
                }

                @Override
                public int getContentLength() {
                    String contentLength = session.getHeaders().get("content-length");
                    return contentLength != null ? Integer.parseInt(contentLength) : 0;
                }

                @Override
                public InputStream getInputStream() throws IOException {
                    return session.getInputStream();
                }
            };

            if (upload == null) {
                initUpFile();
            }

            List<FileItem> items = upload.parseRequest(request);
            List<Map<String, Object>> listFile = new ArrayList<>();

            for (FileItem item : items) {
                if (item.isFormField()) {
                    String value = item.getString("UTF-8");
                    formFields.put(item.getFieldName(), value);
                } else {
                    String fileName = item.getName();
                    String fieldName = item.getFieldName();
                    String contentTypeItem = item.getContentType();
                    long fileSize = item.getSize();

                    File cacheDir = ALSHelper.getCacheDir();
                    File uploadDir = new File(cacheDir, "uploads");
                    if (!uploadDir.exists()) {
                        uploadDir.mkdirs();
                    }

                    File uploadedFile = new File(uploadDir,
                            fileName != null ? fileName : "upload_" + System.currentTimeMillis());
                    item.write(uploadedFile);

                    Map<String, Object> fileInfo = new HashMap<>();
                    fileInfo.put("path", uploadedFile.getAbsolutePath());
                    fileInfo.put("filename", fileName != null ? fileName : "");
                    fileInfo.put("content_type", contentTypeItem != null ? contentTypeItem : "");
                    fileInfo.put("size", fileSize);
                    fileInfo.put("field_name", fieldName);

                    listFile.add(fileInfo);
                }
            }

            result.putAll(formFields);

            for (Map<String, Object> file : listFile) {
                String fieldName = (String) file.get("field_name");
                if (filesInfo.containsKey(fieldName) && filesInfo.get(fieldName) != null) {
                    filesInfo.get(fieldName).add(file);
                } else {
                    List<Map<String, Object>> newList = new ArrayList<>();
                    newList.add(file);
                    filesInfo.put(fieldName, newList);
                }
            }
            for (Map.Entry<String, List<Map<String, Object>>> entry : filesInfo.entrySet()) {
                result.put(entry.getKey(), ALSHelper.listToJSONArray(entry.getValue()).toString());
            }
        } catch (Exception e) {
            e.printStackTrace();
            Log.e(TAG, "Error handling multipart data: " + e.getMessage());
        }
        return result;
    }

    private Response wrapResponse(IHTTPSession session, Response response) {
        String allowHeaders = "Content-Type";
        if (session.getHeaders() != null) {
            String requestHeaders = session.getHeaders().get("access-control-request-headers");
            if (requestHeaders != null) {
                allowHeaders = requestHeaders;
            }
        }
        if (response != null) {
            response.addHeader("Access-Control-Allow-Headers", allowHeaders);
            //[GET, POST, PUT, DELETE, HEAD"]
            response.addHeader("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, HEAD, OPTIONS, TRACE, PATCH, SUBSCRIBE, UNSUBSCRIBE, NOTIFY");
            response.addHeader("Access-Control-Allow-Credentials", "true");
            response.addHeader("Access-Control-Allow-Origin", "*");
            response.addHeader("Access-Control-Max-Age", String.valueOf(42 * 60 * 60));
        }
        return response;
    }

    private String getFileExtensionName(String fileName) {
        return fileName.substring(fileName.lastIndexOf(".") + 1).toLowerCase(Locale.getDefault());
    }

    /**
     * 自定义响应
     */
    public Response customResponse(Exception response) {
        String mimeType = response instanceof CustomResponse ? ((CustomResponse) response).getContentType() : mimeTypes().get("json");
        Response res = Response.newFixedLengthResponse(Status.OK, mimeType, response.getMessage());
        res.addHeader("Access-Control-Allow-Origin", "*");
        res.addHeader("Access-Control-Allow-Methods", "GET, HEAD, POST, PUT, DELETE, OPTIONS, TRACE, PATCH, SUBSCRIBE, UNSUBSCRIBE, NOTIFY");
        res.addHeader("Access-Control-Allow-Credentials", "true");
        res.addHeader("Access-Control-Allow-Headers", "Content-Type, Authorization, X-Requested-With");
        res.addHeader("Access-Control-Max-Age", String.valueOf(42 * 60 * 60));
        if (response instanceof CustomResponse) {
            for (Map.Entry<String, String> entry : ((CustomResponse) response).getHeaders().entrySet()) {
                res.addHeader(entry.getKey(), entry.getValue());
            }
        }
        return res;
    }

    /**
     * 返回JSON响应  字段 code message data
     */
    private Response jsonResponse(int code, String message, Object data) {
        Map<String, Object> result = new HashMap<>();
        result.put(CODE_NAME, code);
        result.put(MESSAGE_NAME, message);
        if (data != null) {
            result.put(DATA_NAME, data);
        }
        return Response.newFixedLengthResponse(
                Status.OK,
                mimeTypes().get("json"),
                ALSHelper.mapToJsonString(result)
        );
    }

    private Response fileResponse(String fileName) {
        AssetManager assetManager = ALSHelper.getAssetManager();
        if (assetManager == null) {
            return fileNotFoundResponse();
        }
        try {
            InputStream stream = assetManager.open(fileName);
            String extension = getFileExtensionName(fileName);
            return Response.newChunkedResponse(Status.OK, mimeTypes().get(extension), stream);
        } catch (IOException e) {
            e.printStackTrace();
            Log.e(TAG, "File not exist. " + fileName);
            return fileError();
        }
    }

    private Response fileNotFoundResponse() {
        return fileNotFoundResponse("Error 404, file not found.");
    }

    private Response fileNotFoundResponse(String error) {
        return Response.newFixedLengthResponse(
                Status.NOT_FOUND,
                MIME_PLAINTEXT,
                error
        );
    }

    public Response success(Object data) {
        return jsonResponse(200, "success", data);
    }

    public Response successEmpty() {
        return jsonResponse(200, "success", null);
    }

    public Response error(String message) {
        return jsonResponse(500, message, null);
    }

    public Response errorPath() {
        return jsonResponse(500, "unknow request path, please check it", null);
    }

    public Response fileSuccess(String fileName) {
        return fileResponse(fileName);
    }

    public Response fileError() {
        return fileNotFoundResponse();
    }

}

File: core/src/main/java/com/android/local/service/core/service/ALSService.kt
package com.android.local.service.core.service

import android.content.res.AssetManager
import android.util.Log
import com.android.local.service.core.ALSHelper
import com.android.local.service.core.i.RequestListener
import org.apache.commons.fileupload.FileUpload
import org.apache.commons.fileupload.RequestContext
import org.apache.commons.fileupload.disk.DiskFileItemFactory
import org.json.JSONArray
import org.json.JSONObject
import org.nanohttpd.protocols.http.IHTTPSession
import org.nanohttpd.protocols.http.NanoHTTPD
import org.nanohttpd.protocols.http.request.Method
import org.nanohttpd.protocols.http.response.Response
import org.nanohttpd.protocols.http.response.Response.newChunkedResponse
import org.nanohttpd.protocols.http.response.Response.newFixedLengthResponse
import org.nanohttpd.protocols.http.response.Status
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.util.Locale


class ALSService(port: Int) : NanoHTTPD(port) {

    private val TAG = this::class.java.simpleName

    private var requestListener: RequestListener? = null
    private var upload: FileUpload? = null

    private fun initUpFile() {
        val factory = DiskFileItemFactory()
        upload = FileUpload(factory)
    }

    fun setRequestListener(listener: RequestListener) {
        this.requestListener = listener
    }

    @Deprecated("Deprecated in Java")
    override fun serve(session: IHTTPSession): Response {
        val method = session.method
        val uri = session.uri
        var params = session.parms
        //Log.d("headers", "${session.headers}")

        var contentType = "text/plain"
        if (session.headers.containsKey("content-type")) {
            contentType = session.headers["content-type"]?.toLowerCase(Locale.getDefault())
                ?: "text/plain"
        }
        if (Method.POST == method) {
            try {
                if (contentType.contains("multipart/form-data")) {// TODO 不太支持
                    params = handleMultipartData(session)
                } else if (contentType.contains("application/json")) {  // 如果是 JSON 请求
                    // 直接从 inputStream 读取 body
                    val body = getRequestBody(session)
                    Log.d("JSON Body", body) // 将 JSON 转换为 Map
                    params = mutableMapOf("json" to body)
                } else {
                    // 传统的表单数据处理
                    session.parseBody(mapOf())
                    params = session.parms
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        // 默认传入的url是以“/”开头的，需要删除掉，否则就变成了绝对路径
        val action = uri?.substring(1) ?: return wrapResponse(session, fileNotFoundResponse())

        Log.d(TAG, "uri = $uri   method = $method   params = $params")
        val response = requestListener?.onRequest(contentType, action, params ?: mapOf())
            ?: error("setRequestListener方法没有设置")
        return wrapResponse(session, response)
    }

    private fun getRequestBody(session: IHTTPSession): String {
        val contentLength = session.headers?.get("content-length")?.toInt() ?: 0
        val buffer = ByteArray(contentLength)
        try {
            session.inputStream.read(buffer, 0, contentLength)
            return String(buffer, Charsets.UTF_8)
        } catch (e: IOException) {
            e.printStackTrace()
            return ""
        }
    }

    private fun handleMultipartData(session: IHTTPSession): Map<String, String> {
        val result = mutableMapOf<String, String>()
        val filesInfo = mutableMapOf<String, MutableList<Map<String, Any>>>()
        val formFields = mutableMapOf<String, String>()
        try {
            val request = object : RequestContext {
                override fun getCharacterEncoding(): String {
                    return session.headers?.get("content-type")?.trim()?.split(";")
                        ?.get(1)?.split("=")?.get(1) ?: "UTF-8"
                }

                override fun getContentType(): String {
                    return session.headers?.get("content-type") ?: "text/plain"
                }

                override fun getContentLength(): Int {
                    return session.headers?.get("content-length")?.toInt() ?: 0
                }

                override fun getInputStream(): InputStream {
                    return session.inputStream
                }
            }
            if (upload == null) {
                initUpFile()
            }
            val items = upload?.parseRequest(request)
            val listFile = arrayListOf<Map<String, Any>>()
            items?.forEach { item ->
                if (item.isFormField) {
                    val value = item.getString("UTF-8")
                    //Log.d("formField:", value)
                    // 普通表单字段
                    formFields[item.fieldName] = value
                } else {//自动识别是文件还是普通字段
                    // 文件上传字段
                    val fileName = item.name
                    val fieldName = item.fieldName
                    val contentType = item.contentType
                    val fileSize = item.size

                    // 保存文件到临时目录
                    val context = ALSHelper.context
                    val cacheDir = context?.externalCacheDir ?: context?.cacheDir
                    val uploadDir = File(cacheDir, "uploads")
                    if (!uploadDir.exists()) {
                        uploadDir.mkdirs()
                    }
                    // 文件名
                    val uploadedFile =
                        File(uploadDir, fileName ?: "upload_${System.currentTimeMillis()}")
                    item.write(uploadedFile)

                    // 创建文件信息映射
                    val fileInfo = mapOf(
                        "path" to uploadedFile.absolutePath,
                        "filename" to (fileName ?: ""),
                        "content_type" to (contentType ?: ""),
                        "size" to fileSize,
                        "field_name" to fieldName
                    )
                    // 实现可多个文件
                    listFile.add(fileInfo)
                    // fieldName 请求端的文件字段
                }
            }
            result.putAll(formFields)

            listFile.forEach {
                val fieldName = it["field_name"] as String
                if (filesInfo.containsKey(fieldName) && filesInfo[fieldName] != null) {
                    filesInfo[fieldName]?.add(it)
                } else {
                    filesInfo[fieldName] = arrayListOf(it)
                }
            }
            filesInfo.forEach {
                result[it.key] = listToJSONArray(it.value).toString()
            }
            //if (listFile.isNotEmpty()) {
                //result["uploaded_files"] = listToJSONArray(listFile).toString()
                //result["uploaded_files"] = mapToJSONObject(filesInfo).toString()
            //}
        } catch (e: Exception) {
            e.printStackTrace()
            Log.e(TAG, "Error handling multipart data: ${e.message}")
        }

        return result
    }


    private fun wrapResponse(session: IHTTPSession, response: Response): Response {
        //下面是跨域的参数
        var allowHeaders = ""
        session.headers?.let {
            val requestHeaders = it["access-control-request-headers"]
            allowHeaders = requestHeaders ?: "Content-Type"
        }
        response.addHeader("Access-Control-Allow-Headers", allowHeaders);
        response.addHeader("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, HEAD");
        response.addHeader("Access-Control-Allow-Credentials", "true");
        response.addHeader("Access-Control-Allow-Origin", "*");
        response.addHeader("Access-Control-Max-Age", "" + 42 * 60 * 60);
        return response
    }

    private fun getFileExtensionName(fileName: String): String {
        return fileName.substring(fileName.lastIndexOf(".") + 1).toLowerCase(Locale.getDefault())
    }

    private fun jsonResponse(code: Int, message: String, data: Any?): Response {
        val result = hashMapOf<String, Any>()
        result["code"] = code
        result["message"] = message
        data?.let {
            result["data"] = it
        }
        return newFixedLengthResponse(
            Status.OK,
            mimeTypes()["json"],
            mapToJsonString(result)
        )
    }

    private fun fileResponse(fileName: String): Response {
        val context = ALSHelper.context ?: return fileNotFoundResponse()
        val assetManager: AssetManager = context.assets
        return try {
            val stream = assetManager.open(fileName)
            val extension: String = getFileExtensionName(fileName)
            newChunkedResponse(Status.OK, mimeTypes()[extension], stream)
        } catch (e: IOException) {
            e.printStackTrace()
            Log.e(TAG, "File not exist. $fileName")
            fileError()
        }
    }

    private fun fileNotFoundResponse(error: String = "Error 404, file not found."): Response {
        return newFixedLengthResponse(
            Status.NOT_FOUND,
            MIME_PLAINTEXT,
            error
        )
    }

    fun success(data: Any) = jsonResponse(code = 200, message = "success", data = data)
    fun successEmpty() = jsonResponse(code = 200, message = "success", data = null)
    fun error(message: String) = jsonResponse(code = 500, message = message, data = null)
    fun errorPath() =
        jsonResponse(code = 500, message = "unknow request path, please check it", data = null)

    fun fileSuccess(fileName: String) = fileResponse(fileName)
    fun fileError() = fileNotFoundResponse()


    /*****************************************************************/


    private fun jsonStringToMap(json: String): MutableMap<String, String> {
        val map = mutableMapOf<String, String>()
        val jsonObject = JSONObject(json)
        for (key in jsonObject.keys()) {
            map[key] = jsonObject.getString(key)
        }
        return map
    }

    private fun jsonToMap(jsonObject: JSONObject): Map<String, Any> {
        val map: MutableMap<String, Any> = HashMap()
        for (key in jsonObject.keys()) {
            val value = jsonObject[key]
            if (value is JSONObject) {
                map[key] = jsonToMap(value)
            } else if (value is JSONArray) {
                map[key] = jsonArrayToList(value)
            } else {
                map[key] = value
            }
        }
        return map
    }

    private fun jsonArrayToList(jsonArray: JSONArray): List<Any> {
        val list: MutableList<Any> = ArrayList()
        for (i in 0 until jsonArray.length()) {
            val value = jsonArray[i]
            if (value is JSONObject) {
                list.add(jsonToMap(value))
            } else if (value is JSONArray) {
                list.add(jsonArrayToList(value))
            } else {
                list.add(value)
            }
        }
        return list
    }

    private fun mapToJsonString(map: Map<String, Any>): String {
        // 方法1: 直接使用Map构造JSONObject
        val jsonObject = JSONObject(map)
        return jsonObject.toString()
    }

    private fun mapToJsonString2(map: Map<String, Any>): String {
        // 方法2: 手动遍历添加
        val jsonObject2 = JSONObject()
        for ((key, value) in map.entries) {
            jsonObject2.put(key, value)
        }
        return jsonObject2.toString()
    }

    private fun mapToJSONObject(map: Map<String, Any>): JSONObject {
        val jsonObject = JSONObject()
        for ((key, value) in map) {
            if (value is Map<*, *>) {
                // 递归处理嵌套Map
                jsonObject.put(key, mapToJSONObject(value as Map<String, Any>))
            } else if (value is List<*>) {
                // 处理List
                jsonObject.put(key, listToJSONArray(value))
            } else {
                // 处理基本类型
                jsonObject.put(key, value)
            }
        }
        return jsonObject
    }

    private fun listToJSONArray(list: List<*>): JSONArray {
        val jsonArray = JSONArray()
        for (item in list) {
            if (item is Map<*, *>) {
                jsonArray.put(mapToJSONObject(item as Map<String, Any>))
            } else if (item is List<*>) {
                jsonArray.put(listToJSONArray(item))
            } else {
                jsonArray.put(item)
            }
        }
        return jsonArray
    }
}